[Java] Getting started with Kyma guide (#682)
SAP/cloud-sdk: docs/java/environments/sap-btp-kyma-environment.mdx

## Getting Started with the SAP Cloud SDK on Kyma

This detailed guide will help get your SAP Cloud SDK Java application up and running in Kyma.
You can also use this guide to migrate your existing application to Kyma.

### Prerequisites

To follow this guide you will need:

- An [SAP BTP Kyma environment](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/09dd313bf6644250a14f8f38c3d644c0.html?locale=en-US) instance
- [Docker](https://www.docker.com/) and a publicly reachable Docker repository

> - A [Spring Boot](https://spring.io/projects/spring-boot) Application using the SAP Cloud SDK

Reviewer: Maybe we should add a note/tip that it is not a strict requirement for applications on Kyma to be based on Spring Boot. 
This guide just happens to use a Spring Boot application and, therefore, following it will be easiest when also having such an application at hand.

Check out the details below in case you are uncertain about any of the prerequisites.

<details>

>   <summary>Kyma Environment</summary>

Reviewer: I think we should add a few sentences about how to configure access from your local CLI to the Kyma cluster.

In your subaccount in SAP BTP cockpit, create a [Kyma Environment instance](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/09dd313bf6644250a14f8f38c3d644c0.html?locale=en-US).
Make sure that you select an appropriate [plan](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/befe01d5d8864e59bf847fa5a5f3d669.html?locale=en-US#azure) and also assign the necessary [roles](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/148ae38b7d6f4e61bbb696bbfb3996b2.html?locale=en-US) to users.

:::note

Creating a Kyma environment instance can take up to 30 minutes.

:::

</details>

<details>
  <summary>Docker</summary>

This guide assumes you have [Docker](https://www.docker.com/) installed on your local machine.

Furthermore, you need a **Docker repository** where you can store images.
The repository needs to be publicly accessible in order for the cluster to access and download the Docker image we are going to create.

In case you don't have such a repository yet we recommend either:

- [Docker Hub](https://hub.docker.com/)

> - [Artifactory DMZ](https://common.repositories.cloud.sap/ui/) (for SAP internal developers)

Reviewer 2: Actually the link is accessible but behind a login, so it's okay I'd say 👍🏻

If you don't have an application already you can comfortably [create one from our archetypes](../getting-started).

</details>

### Containerize the Application

To run on Kyma the application needs to be shipped in a container.
For this guide we will be using Docker.

Create a `Dockerfile` in the project root directory:

```dockerfile {2}
FROM openjdk:8u312-jdk-slim
ARG JAR_FILE=application/target/*.jar
COPY ${JAR_FILE} app.jar
ENTRYPOINT ["java","-jar","/app.jar"]
```

If needed, update the `JAR_FILE` to point to your JAR file.

:::tip
You can find more information on how to containerize Spring Boot applications in [this guide](https://spring.io/guides/gs/spring-boot-docker/) (in particular, check the _Containerize It_ section).
:::

Compile and push the image by running:

```bash
docker build -t <your-repo>/<your-image-name> .
docker push <your-repo>/<your-image-name>
```

:::tip
In case you are facing authorization issues when pushing to your repository refer to the dedicated section under [Prerequisites](#prerequisites).
:::

### Create a Kyma Deployment

1. Create a new YAML file:

   ```yml title="deployment.yml" {17,31}

>    apiVersion: apps/v1
>    kind: Deployment
>    metadata:
>    name: my-app-deployment
>    spec:
>    replicas: 1
>    selector:
>      matchLabels:
>        app: my-app
>    template:
>      metadata:
>        labels:
>          app: my-app
>      spec:
>        containers:
>          # Configure the docker image you just pushed to your repository here
>          - image: <name-of-the-image>
>            name: my-app
>            imagePullPolicy: Always
>            resources:
>              requests:
>                memory: '1.5Gi'
>                cpu: '500m'
>              limits:
>                memory: '2Gi'
>                cpu: '1000m'
>            # Volume mounts needed for injecting SAP BTP service credentials
>            volumeMounts:
>        imagePullSecrets:
>          # In case your repository requires a login, reference your secret here
>          - name: <your-secret-for-docker-login>
>        volumes:
> 
>    ----
>    apiVersion: v1
>    kind: Service
>    metadata:
>      labels:
>        app: my-app
>      name:  my-app
>    spec:
>      type: NodePort
>      ports:
>        - port: 8080
>          targetPort: 8080
>          protocol: TCP
>      selector:
>        app: my-app

Reviewer: Indentation is slightly off.

```suggestion
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: my-app-deployment
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: my-app
      template:
        metadata:
          labels:
            app: my-app
        spec:
          containers:
            # Configure the docker image you just pushed to your repository here
            - image: <name-of-the-image>
              name: my-app
              imagePullPolicy: Always
              resources:
                requests:
                  memory: '1.5Gi'
                  cpu: '500m'
                limits:
                  memory: '2Gi'
                  cpu: '1000m'
              # Volume mounts needed for injecting SAP BTP service credentials
              volumeMounts:
          imagePullSecrets:
            # In case your repository requires a login, reference your secret here
            - name: <your-secret-for-docker-login>
          volumes:
    ----
    apiVersion: v1
    kind: Service
    metadata:
      labels:
        app: my-app
      name:  my-app
    spec:
      type: NodePort
      ports:
        - port: 8080
          targetPort: 8080
          protocol: TCP
      selector:
        app: my-app
```

2. Install the configuration via `kubectl apply -f path/to/deployment/file/deployment.yml`.

3. Monitor the status of the deployment by running: `kubectl get deployment my-app-deployment`.

Eventually, you should see an output similar to:

```bash
kubectl get deployment my-app-deployment

NAME                  READY   UP-TO-DATE   AVAILABLE   AGE
my-app-deployment     1/1     1            1           15s
````

:::tip
In case something went wrong use `kubectl describe` together with `deployment` or `pod` to get more information about the status of your application.
:::

### Access Your Application inside the cluster

To access your application from inside the cluster find the name of the application pod by listing out all available pods:

```bash
kubectl get pods
```

Now use the name of the application pod and open a shell in this pod using:

```bash
kubectl exec --stdin --tty [pod-name] -- /bin/sh
```

Install [`curl`](https://curl.se/) on your pod depending on the Linux distribution of your pod and access the application by accessing an endpoint using `curl`.

In case you started with an SAP Cloud SDK Archetype, you can test the `hello` endpoint:

```
curl localhost:8080/hello

> ```

Reviewer: Not sure if that is what we want to recommend....
How about recommended/demonstrating how to properly use `kubectl port-forward` instead?
That way, users don't have to install `curl` each time they deploy a new version of their image.

We have observed that the property `plan` is missing in the XSUAA secret created in the previous section.
This property, however, is required by the SAP Cloud SDK and it's absence causes runtime errors in the application while trying to read this service binding.

To fix this issue you need to edit the secret so that it contains the `plan` property.
The easiest way, when you are already using the CLI, is by using the `kubectl edit` command:

```bash
kubectl edit secrets my-xsuaa-secret
```

In there, you can now add a `plan` property with one of the following base64 encoded values:

| plan          | base64 encoded value |
| ------------- | -------------------- |
| `application` | `YXBwbGljYXRpb24K`   |
| `broker`      | `YnJva2VyCg==`       |
| `space`       | `c3BhY2UK`           |
| `apiaccess`   | `YXBpYWNjZXNzCg==`   |

The resulting service binding can now be consumed with the SAP Cloud SDK.

### Access Your Application outside the cluster

We will configure and deploy an [Approuter](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/01c5f9ba7d6847aaaf069d153b981b51.html?locale=en-US) so that only authenticated users can access our application.

> We created a simple application that uses the [@sap/approuter](https://www.npmjs.com/package/@sap/approuter) and starts it for this purpose.

Reviewer: ```suggestion
You will need to create a simple application that uses the [@sap/approuter](https://www.npmjs.com/package/@sap/approuter) and starts it for this purpose.
```

         nodePort: 31002
     selector:
       app: my-approuter
   ```

3. Also adapt the `xs-app.json` file:

```json{5,7,8}
{
    "welcomeFile": "/web-pages/index.html",
    "routes": [
        {
            "source": "<path-for-backend-endpoints>",
            "target": "$1",
            "destination": "<name-for-the-backend-app>",
            "identityProvider": "<optional-identity-provider-id>",
            "csrfProtection": false
        }
    ]
}
```

4. Follow steps similar to steps 2-3 in [Create a Kyma Deployment section](#create-a-kyma-deployment) to deploy the approuter and check if the deployment is successful.

5. Update the already created XSUAA service instance with few parameters like `xsappname` and `oauth2-configuration`.
   This is required to provide authenticated access to backend application via the approuter.

```yml title="xsuaa-service.yaml"{9-15}
---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceInstance
metadata:
  name: xsuaa-service
spec:
  clusterServiceClassExternalName: xsuaa
  clusterServicePlanExternalName: <your-desired-plan>
  parameters:
    xsappname: <your-backend-app-name>
    oauth2-configuration:
      allowedproviders:

>         - sap.custom

Reviewer: ```suggestion
        - <your-idp-key>
```

## Getting Started with the SAP Cloud SDK on Kyma

This detailed guide will help get your SAP Cloud SDK Java application up and running in Kyma.
You can also use this guide to migrate your existing application to Kyma.

### Prerequisites

To follow this guide you will need:

- An [SAP BTP Kyma environment](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/09dd313bf6644250a14f8f38c3d644c0.html?locale=en-US) instance

> - [Docker](https://www.docker.com/) and a publicly reachable Docker repository

Reviewer 3: (question)
What do I need Docker for? Locally(?)

Author: As we are building container images and pushing it to a repository, I thought this is required.
Or is there an easier way to do this?

If not, please refer to these guides on getting a [Global Account](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/d61c2819034b48e68145c45c36acba6e.html?locale=en-US) and creating a [subaccount](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/05280a123d3044ae97457a25b3013918.html?locale=en-US)

In your subaccount in SAP BTP cockpit, create a [Kyma Environment instance](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/09dd313bf6644250a14f8f38c3d644c0.html?locale=en-US).
Make sure that you select an appropriate [plan](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/befe01d5d8864e59bf847fa5a5f3d669.html?locale=en-US#azure) and also assign the necessary [roles](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/148ae38b7d6f4e61bbb696bbfb3996b2.html?locale=en-US) to users.

:::note

Creating a Kyma environment instance can take up to 30 minutes.

:::

</details>

<details>
  <summary>Docker</summary>

This guide assumes you have [Docker](https://www.docker.com/) installed on your local machine.

> Furthermore, you need a **Docker repository** where you can store images.

Reviewer 3: (question)
Does the Kyma Environment not have a local repository?
Can't I deploy custom images, i.e. plain `Dockerfile` files?

Reviewer 2: AFAIK no, similar to how BTP / CF doesn't have it's own Maven Repository...

Author: I didn't understand the question fully? Can you please elaborate?
In the our sample Kyma application, Andras seems to have pushed the images to his private repository, I am not sure if he created that in DockerHub.
For our K8s application, we used to push the docker images to Jfrog Artifactory.

Reviewer 3: > Does the Kyma Environment not have a local repository?
> Can't I deploy custom images, i.e. plain `Dockerfile` files?

I thought if the server already has a Docker daemon running, why not allow building _Docker images_ from a _Dockerfile_ locally. But considering this may exceed the platform-as-a-service expectation, I'm totally fine with a **"no"**.

Reviewer 2: the dockerfile is just the instructions how to build the image, it doesn't contain the artifacts.. similarly you can't deploy your app by uploading pom.xml files

- [Artifactory DMZ](https://common.repositories.cloud.sap/ui/) (for SAP internal developers)

Access to images in a repository may be limited to authenticated and/or authorized users, depending on your configuration.

Make sure you are logged in to your repository on your local machine by running:

```bash
docker login (your-repo) --username=(your-username)
```

And check your configuration which is usually located under `(your-home-directory)/.docker/config.json`.

:::tip
In case AuthN/AuthZ is required to download images make sure you have a secret configured in your cluster

```bash
kubectl create secret docker-registry (name-of-the-secret) --docker-username=(username) --docker-password=(API-token) --docker-server=(your-repo)
```

:::

</details>

<details>
  <summary>Application using the SAP Cloud SDK</summary>

If you don't have an application already you can comfortably [create one from our archetypes](../getting-started).

</details>

### Containerize the Application

To run on Kyma the application needs to be shipped in a container.
For this guide we will be using Docker.

Create a `Dockerfile` in the project root directory:

```dockerfile {2}

> FROM openjdk:8u312-jdk-slim

Reviewer 3: (suggestion)
I would say, let's use a wildcard version(?)
I.e. not necessarily concrete update `312` which may be out-of-date tomorrow

Reviewer 4: As far as I understood docker it makes sense in a productive environment to be as specific as possible, in theory even qualifying the hash. That way the build is 100% reproducible and doesn't depend on the date it was built.

However, for documentation purposes I would just mention version 8 here, as the patch/update version doesn't really matter.

Author: Done.

ARG JAR_FILE=application/target/*.jar
COPY ${JAR_FILE} app.jar
ENTRYPOINT ["java","-jar","/app.jar"]
```

If needed, update the `JAR_FILE` to point to your JAR file.

:::tip

> You can find more information on how to containerize Spring Boot applications in [this guide](https://spring.io/guides/gs/spring-boot-docker/) (in particular, check the _Containerize It_ section).

Reviewer 3: (question) 
Is it required for Kyma to only work with docker images?
I cannot use the CLI based approach...?
```
docker run --rm -v "$PWD":/usr/src/myapp -w /usr/src/myapp openjdk:8u312-jdk-slim java -jar app.jar
```
because there's no option to "upload" an `app.jar` artifact, right?

Reviewer 2: For µServices yes, we need a container that Kubernetes can run. For server less functions there seems to be the option to just [provide a GH repo](https://kyma-project.io/docs/kyma/latest/03-tutorials/00-serverless/svls-02-create-git-function/).

:::

### Create a Kyma Deployment

1. Create a new YAML file:
   <!-- vale off -->
   ```yml title="deployment.yml" {17,31}
   apiVersion: apps/v1
   kind: Deployment
   metadata:
   name: my-app-deployment
   spec:
   replicas: 1
   selector:
     matchLabels:
       app: my-app
   template:
     metadata:
       labels:
         app: my-app
     spec:
       containers:
         # Configure the docker image you just pushed to your repository here
         - image: <name-of-the-image>
           name: my-app
           imagePullPolicy: Always
           resources:
             requests:
               memory: '1.5Gi'
               cpu: '500m'
             limits:
               memory: '2Gi'
               cpu: '1000m'
           # Volume mounts needed for injecting SAP BTP service credentials
           volumeMounts:
       imagePullSecrets:
         # In case your repository requires a login, reference your secret here
         - name: <your-secret-for-docker-login>
       volumes:
   ```

---

apiVersion: v1
kind: Service
metadata:
labels:
app: my-app
name: my-app
spec:
type: NodePort
ports: - port: 8080
targetPort: 8080
protocol: TCP
selector:

> app: my-app

Reviewer 3: (issue)
Missing Markdown format

> ````

Reviewer 3: (issue)
Is this proper Markdown with four ` ?

Author: Fixed it.

3. Monitor the status of the deployment by running: `kubectl get deployment my-app-deployment`.

Eventually, you should see an output similar to:

```bash
kubectl get deployment my-app-deployment

NAME                  READY   UP-TO-DATE   AVAILABLE   AGE
my-app-deployment     1/1     1            1           15s
````

:::tip
In case something went wrong use `kubectl describe` together with `deployment` or `pod` to get more information about the status of your application.
:::

### Access Your Application inside the cluster

To access your application from inside the cluster find the name of the application pod by listing out all available pods:

```bash
kubectl get pods
```

Now use the name of the application pod and open a shell in this pod using:

```bash
kubectl exec --stdin --tty [pod-name] -- /bin/sh
```

Install [`curl`](https://curl.se/) on your pod depending on the Linux distribution of your pod and access the application by accessing an endpoint using `curl`.

In case you started with an SAP Cloud SDK Archetype, you can test the `hello` endpoint:

```
curl localhost:8080/hello
```

### Bind SAP BTP Services to the Application

> The SAP Business Technology Platform offers various services that can be used by applications.

Reviewer 2: Here it is probably worthwhile to mention the UI-variant first and show a screenshot, since this is a noteable difference to Gardener. For the CLI approach we could then consider using a `<details>` section

Reviewer: Done 👍

We have observed that the property `plan` is missing in the XSUAA secret created in the previous section.
This property, however, is required by the SAP Cloud SDK and it's absence causes runtime errors in the application while trying to read this service binding.

To fix this issue you need to edit the secret so that it contains the `plan` property.
The easiest way, when you are already using the CLI, is by using the `kubectl edit` command:

```bash
kubectl edit secrets my-xsuaa-secret
```

In there, you can now add a `plan` property with one of the following base64 encoded values:

| plan          | base64 encoded value |
| ------------- | -------------------- |
| `application` | `YXBwbGljYXRpb24K`   |
| `broker`      | `YnJva2VyCg==`       |
| `space`       | `c3BhY2UK`           |
| `apiaccess`   | `YXBpYWNjZXNzCg==`   |

The resulting service binding can now be consumed with the SAP Cloud SDK.

### Access Your Application outside the cluster

We will configure and deploy an [Approuter](https://help.sap.com/products/BTP/65de2977205c403bbc107264b8eccf4b/01c5f9ba7d6847aaaf069d153b981b51.html?locale=en-US) so that only authenticated users can access our application.

> We used a slimmed down version of this [repository](https://github.wdf.sap.corp/xs2/approuter.js) for this purpose.

Author: The repo I refer to for approuter is SAP internal link, is that fine?

Reviewer 3: I'm afraid that won't be okay

Reviewer 4: I think this link should be good enough: https://www.npmjs.com/package/@sap/approuter

Author: I reworded it to: 
We created a simple application that uses the [@sap/approuter](https://www.npmjs.com/package/@sap/approuter) and starts it for this purpose.

                 value: >
                   [
                     {
                       "name": "<name-for-the-backend-app>",
                       "url": "<URL-to-access-backend-app>",
                       "forwardAuthToken": true
                     }
                   ]
               - name: TENANT_HOST_PATTERN
                 value: <URL-pattern-for-outside-cluster-access>
             volumeMounts:
               - name: my-xsuaa-service-binding-volume
                 mountPath: '/etc/secrets/sapcp/xsuaa/my-xsuaa-service-binding'
         volumes:
           - name: my-xsuaa-service-binding-volume
             secret:
               secretName: my-xsuaa-secret
         restartPolicy: Always

   ---
   apiVersion: v1
   kind: Service
   metadata:
     labels:
       app: my-approuter
     name: my-approuter
   spec:
     type: NodePort
     ports:
       - port: 5000
         targetPort: 5000
         protocol: TCP
         nodePort: 31002
     selector:
       app: my-approuter
   ```

3. Also adapt the `xs-app.json` file:

```json{5,7,8}
{
    "welcomeFile": "/web-pages/index.html",
    "routes": [
        {
            "source": "<path-for-backend-endpoints>",

>             "target": "$1",

Reviewer 3: (issue)
is this a valid target?
Looks like a regular expression match indicator. Were there braces in the `"source"` field?

Author: Yes, the [file](https://github.tools.sap/cloudsdk/kyma-approuter/blob/main/xs-app.json) I referred to had "/backend-app/(.*)" as source value. 
I have to read and check then on what could be a valid value.

Reviewer 3: Why not...?
```
"source": "/(.*)",
```

or alternatively...?
```
"source": "/<path>/(.*)",
```

            "destination": "<name-for-the-backend-app>",

>             "identityProvider": "<identity-provider-url>",

Reviewer 3: ```suggestion
            "identityProvider": "<optional-identity-provider-id>",
```

```yml title="xsuaa-service.yaml"{9-15}
---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceInstance
metadata:
  name: xsuaa-service
spec:
  clusterServiceClassExternalName: xsuaa
  clusterServicePlanExternalName: <your-desired-plan>
  parameters:
    xsappname: <your-backend-app-name>
    oauth2-configuration:
      allowedproviders:
        - sap.custom
      redirect-uris:
        - <URL-pattern-for-outside-cluster-access>
---
apiVersion: servicecatalog.k8s.io/v1beta1
kind: ServiceBinding
metadata:
  name: my-xsuaa-service-binding
spec:
  instanceRef:
    name: xsuaa-service
  secretName: my-xsuaa-secret
```

6. Update the configuration via `kubectl apply -f xsuaa-service.yaml`.

7. Make the approuter accessbile outside the cluster by creating an API rule:

```yml title="create-api-rule.yml"{4,15}
apiVersion: gateway.kyma-project.io/v1alpha1
kind: APIRule
metadata:
  name: <your-rule-name>
spec:
  gateway: kyma-gateway.kyma-system.svc.cluster.local
  rules:
    - accessStrategies:
        - config: {}
          handler: noop
      methods:
        - GET
      path: /.*
  service:

>     host: s4sdk-k8s-kyma-q2wk189v.<URL-pattern-for-outside-cluster-access>

Author: I am not sure how this value `s4sdk-k8s-kyma-q2wk189v` was picked in the host? Any suggestions on what could be the apt placeholder here?

Reviewer: AFAIK this is just a random suffix that is added automatically when creating the Kyma environment (since I guess it would be possible to add multiple Kyma environments to the same Subaccount?).

In general, the entire prefix `s4sdk-k8s-kyma-q2wk189v` shouldn't be mentioned here IMO since this is very specific to our internal infrastructure.
Instead, a placeholder like `<subaccount-prefix>` might be better.
